### PR TITLE
Fix two bugs in auto-transactions.

### PR DESCRIPTION
**Pull request by petermattis**
The first bug was obvious and not currently harmful: we were creating
the plan node inside the retryable function passed to DB.Txn(), but then
executing it outside. This happened to work because mutating plans
currently do there work during plan creation and read-only
plans (e.g. SELECT) were unsafely using a committed transaction record.

The second bug was much more subtle. When retrying an
auto-transaction (i.e. an SQL statement executed outside of a BEGIN/END
block) we were passing the same parsed statement for execution. But
statement execution (i.e. SELECT) modifies the parsed statement,
performing normalization (ok) and resolving qualified names. The name
resolution (scanNode.resolveQNames) was not properly handling a parsed
statement for which name resolution had already been performed. The
result was that qualified names were not properly looked up on retries,
using values determined during the first attempt at the transaction.

Added -transfer-style flag to sql_bank and a new optimized SQL style
that moves money in a single UPDATE statement. Removed the timing
warnings now that the egregious problems with statement execution have
been fixed. The remaining warnings were mainly just annoyance.

### CODE REVIEW

File: sql/scan.go [hunk 1]
```diff
@@ -764,6 +764,16 @@ func (v *qnameVisitor) Visit(expr parser.Expr, pre bool) (parser.Visitor, parser
...
 	switch t := expr.(type) {
+	case *qvalue:
+		// We will encounter a qvalue in the expression during retry of an
+		// auto-transaction. When that happens, we've already gone through
+		// qualified name normalization and lookup, we just need to hook the qvalue
+		// up to the scanNode.
+		//
+		// TODO(pmattis): Should we be more careful about ensuring that the various
+		// statement implementations do not modify the AST nodes they are passed?
+		return v, v.getQVal(t.col)
```
Comment (petermattis) [inline sql/scan.go]: This clearly needs a test. I'll work one up before merging.

File: sql/server.go [hunk 2]
```diff
@@ -116,67 +116,78 @@ func (s server) execStmt(stmt parser.Statement, params parameters, planMaker *pl
...
-			plan, err = planMaker.makePlan(stmt)
-			planMaker.txn = nil
+
+	// Create a function which both makes and executes the plan, populating
+	// result.
+	//
+	// TODO(pmattis): Should this be a separate function? Perhaps we should move
+	// some of the common code back out into execStmts and have execStmt contain
+	// only the body of this closure.
+	f := func() error {
```
Comment (petermattis) [inline sql/server.go]: I'm not sure how to test this change. The previous code was working, though accidentally.

**Comment by tamird** [comment]
LGTM

File: sql/scan.go (hunk 1 above)
Comment (tschottdorf) [inline sql/scan.go]: I'm not sure about the large scale of things here, but maybe the way we're retrying is out of order? It looks like we're taking work from one iteration of the retry into the next, which I don't think we should be doing. Instead, each retry should start from zero.

Comment (petermattis) [inline sql/scan.go]: Yes, we're taking work for one loop iteration and using it in the next. This isn't intentional, just an artifact of the way the code was written. As the `TODO` mentions, it would probably be better for this not to happen.

File: examples/sql_bank/main.go [hunk 3]
```diff
@@ -36,7 +37,7 @@ import (
 var maxTransfer = flag.Int("max-transfer", 999, "Maximum amount to transfer in one transaction.")
 var numAccounts = flag.Int("num-accounts", 999, "Number of accounts.")
 var concurrency = flag.Int("concurrency", 5, "Number of concurrent actors moving money.")
-var aggregate = flag.Bool("aggregate", true, "Use aggregate function to verify conservation of money.")
+var transferStyle = flag.Int("transfer-style", 0, "0: single UPDATE, 1: transactional")
```
Comment (bdarnell) [inline examples/sql_bank/main.go]: It would be friendlier to use strings for this instead of magic numbers.

File: examples/sql_bank/main.go [hunk 4]
```diff
@@ -49,76 +50,80 @@ func moveMoney(db *sql.DB) {
...
-			}
-			if err = tx.Rollback(); err != nil {
-				log.Fatal(err)
+
+		switch *transferStyle {
+		case 0:
+			update := `
+UPDATE bank.accounts
+  SET balance = CASE id WHEN $1 THEN balance-$3 WHEN $2 THEN balance+$3 END
+  WHERE id IN ($1, $2)
```
Comment (bdarnell) [inline examples/sql_bank/main.go]: This needs a guard against the "from" account going negative. Actually, it looks like the multi-statement version doesn't have this either; it got lost in the transition from `kv_bank`.

File: examples/sql_bank/main.go [hunk 5]
```diff
@@ -49,76 +50,80 @@ func moveMoney(db *sql.DB) {
...
-			default:
-				panic(fmt.Sprintf("got unexpected account %d", id))
-			}
-		}
-		startTime = time.Now()
-		update := fmt.Sprintf("UPDATE accounts SET balance=%d WHERE id=%d", fromBalance-amount, from)
-		if _, err = tx.Exec(update); err != nil {
-			if log.V(1) {
-				log.Warning(err)
+			query := fmt.Sprintf("SELECT id, balance FROM accounts WHERE id IN (%d, %d)", from, to)
```
Comment (bdarnell) [inline examples/sql_bank/main.go]: Not the fault of this change, but we should use placeholders instead of sprintf for all query construction.

**Comment by bdarnell** [comment]
LGTM

File: examples/sql_bank/main.go (hunk 5 above)
Comment (petermattis) [inline examples/sql_bank/main.go]: Yeah, agreed. I've been fixing that when I've noticed it.

File: examples/sql_bank/main.go (hunk 3 above)
Comment (petermattis) [inline examples/sql_bank/main.go]: Ok. Changed to `"single-stmt"` and `"txn"`.

File: examples/sql_bank/main.go (hunk 4 above)
Comment (petermattis) [inline examples/sql_bank/main.go]: Yeah, I noticed that in testing. I'll fix.

Comment (petermattis) [inline examples/sql_bank/main.go]: We don't currently expand placeholders in sub-selects correctly. Should be easy to fix, but I'm going to do so in another PR and leave a TODO here.

File: sql/scan.go (hunk 1 above)
Comment (petermattis) [inline sql/scan.go]: `TestRetryResolveQNames` added to `scan_test.go`. Fails without this fix and succeeds with it.

**Comment by tamird** [comment]
LGTM
